src/games/strategy/triplea/ui/screen/IDrawable.java

```diff
@@ -180,6 +178,59 @@ public void draw(final Rectangle bounds, final GameData data, final Graphics2D g
     }
   }
 
+  /**
+   * Find the best rectangle inside the territory to place the name in. Finds the rectangle
+   * that can fit the name, that is the closest to the vertical center, and has a large width at
+   * that location. If there isn't any rectangles that can fit the name then default back to the
+   * bounding rectangle.
+   */
+  private Rectangle getBestTerritoryNameRect(final MapData mapData, final Territory territory,
+      final FontMetrics fontMetrics) {
+
+    // Find bounding rectangle and use to create parameters for a grid across the territory
+    final Rectangle territoryBounds = mapData.getBoundingRect(territory);
+    Rectangle result = territoryBounds;
+    final int maxX = territoryBounds.x + territoryBounds.width;
+    final int maxY = territoryBounds.y + territoryBounds.height;
+    final int centerY = territoryBounds.y + territoryBounds.height / 2;
+    final int minWidth = fontMetrics.stringWidth(territory.getName());
+    final int increment = fontMetrics.getAscent();
+    int maxScore = 0;
+
+    // Loop through the grid moving the starting point and determining max width at that point
+    for (int x = territoryBounds.x; x < maxX - increment; x += increment) {
+      for (int y = territoryBounds.y; y < maxY - increment; y += increment) {
+        for (int endX = maxX; endX > x; endX -= increment) {
+          final Rectangle current = new Rectangle(x, y, endX - x, increment);
+
+          // Ranges from 0 when at very top or bottom of territory to height/2 when at vertical center
+          final int verticalDistanceFromEdge = territoryBounds.height / 2 - Math.abs(centerY - y - increment);
+
+          // Score rectangle based on how close to vertical center and territory width at location
+          final int score = verticalDistanceFromEdge * current.width;
+          if (current.width > minWidth && score > maxScore) {
+
+            // Check to make sure rectangle is contained in the territory
+            boolean isContained = false;
```

Comment:
This is a good place to insert a helper function.
`boolean isContained = isRectangleContainedInTerritory(current, polygons);`